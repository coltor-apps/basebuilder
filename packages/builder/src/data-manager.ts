import {
  createSubscriptionManager,
  type Subscribe,
} from "./subscription-manager";

export function createDataManager<TData>(initialData: TData): {
  getData: () => TData;
  setData: (data: TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let data: TData = initialData;

  function getData(): TData {
    return data;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(newData) {
      data = newData;

      notify(data);

      return data;
    },
  };
}
