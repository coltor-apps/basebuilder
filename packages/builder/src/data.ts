import { createSubscriptionManager, type Subscribe } from "./subscription";

export function createDataManager<TData>(data: TData): {
  getData: () => TData;
  setData: (setter: (oldData: TData) => TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let localData: TData = data;

  function getData(): TData {
    return localData;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(setter) {
      localData = setter(getData());

      notify(localData);

      return localData;
    },
  };
}
