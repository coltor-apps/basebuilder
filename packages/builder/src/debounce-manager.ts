export type DebounceManager<TResult> = {
  handle(
    key: string,
    callback: () => Promise<TResult> | TResult,
    fallback: () => TResult,
  ): Promise<TResult>;
  get(key: string): Date | undefined;
};

function isTimestampCurrent(
  timestamps: Map<string, Date>,
  key: string,
  timestamp: Date,
): boolean {
  return timestamps.get(key)?.getTime() !== timestamp.getTime();
}

export function createDebounceManager<TResult>(): DebounceManager<TResult> {
  const timestamps: Map<string, Date> = new Map();

  return {
    async handle(key, callback, fallback) {
      const currentTimestamp = new Date();

      timestamps.set(key, currentTimestamp);

      try {
        const result = await callback();

        if (isTimestampCurrent(timestamps, key, currentTimestamp)) {
          return fallback();
        }

        return result;
      } catch (error) {
        throw error;
      } finally {
        if (isTimestampCurrent(timestamps, key, currentTimestamp)) {
          timestamps.delete(key);
        }
      }
    },
    get(key) {
      return timestamps.get(key);
    },
  };
}
