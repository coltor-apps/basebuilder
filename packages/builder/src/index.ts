export { createBuilder } from "./builder";
export type { Builder, BuilderEntities } from "./builder";

export { createEntity } from "./entity";
export type { Entity } from "./entity";

export { createInput } from "./input";
export type { Input } from "./input";

export { SchemaValidationError as EntityValidationError, createStore } from "./store";
export type {
  Store,
  StoreData,
  StoreEntity,
  SchemaValidationErrorCode as EntityValidationErrorCause,
} from "./store";

export { getEntitiesNamesExcept } from "./utils";
