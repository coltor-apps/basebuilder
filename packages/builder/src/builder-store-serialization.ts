import { type Builder } from "./builder";
import {
  type BuilderStoreData,
  type BuilderStoreEntity,
  type SerializedBuilderStoreData,
} from "./builder-store";

export function serializeBuilderStoreData<TBuilder extends Builder>(
  data: BuilderStoreData<TBuilder>,
): SerializedBuilderStoreData<TBuilder> {
  const newEntities: SerializedBuilderStoreData<TBuilder>["schema"]["entities"] =
    {};

  for (const [id, entity] of data.schema.entities) {
    const { children, ...entityData } = entity;

    newEntities[id] = {
      ...entityData,
      ...(children ? { children: Array.from(children) } : {}),
      inputs:
        entityData.inputs as unknown as SerializedBuilderStoreData<TBuilder>["schema"]["entities"][string]["inputs"],
    };
  }

  return {
    schema: {
      root: Array.from(data.schema.root),
      entities: newEntities,
    },
    entitiesInputsErrors: Object.fromEntries(data.entitiesInputsErrors),
  };
}

export function deserializeBuilderStoreData<TBuilder extends Builder>(
  data: SerializedBuilderStoreData<TBuilder>,
): BuilderStoreData<TBuilder> {
  return {
    schema: {
      entities: new Map(
        Object.entries(data.schema.entities).map(([id, entity]) => [
          id,
          {
            ...entity,
            ...(entity.children ? { children: new Set(entity.children) } : {}),
            inputs:
              entity.inputs as unknown as BuilderStoreEntity<TBuilder>["inputs"],
          },
        ]),
      ),
      root: new Set(data.schema.root),
    },
    entitiesInputsErrors: new Map(Object.entries(data.entitiesInputsErrors)),
  };
}
