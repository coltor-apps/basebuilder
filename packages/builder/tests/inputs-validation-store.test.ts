import { describe, expect, it } from "vitest";
import { z } from "zod";

import {
  createBuilder,
  createEntity,
  createInput,
  createSchemaStore,
} from "../src";
import { createInputsValidationStore } from "../src/inputs-validation-store";

describe("inputs validation store", () => {
  it("can be created", () => {
    const builder = createBuilder({
      entities: [],
    });

    const schemaStore = createSchemaStore({ builder });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    expect(inputsValidationStore).toMatchSnapshot();
  });

  it("can validate a single entity input", async () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return z.string().parse(value);
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {
              // @ts-expect-error Intentional wrong data type
              label: 1,
            },
          },
          "51324b32-adc3-4d17-a90e-66b5453935bd": {
            type: "test",
            // @ts-expect-error Intentional wrong data type
            inputs: {},
          },
        },
        root: [
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
          "51324b32-adc3-4d17-a90e-66b5453935bd",
        ],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    await expect(
      inputsValidationStore.validateEntityInput(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "label",
      ),
    ).resolves.toEqual(undefined);

    await expect(
      inputsValidationStore.validateEntityInput(
        "51324b32-adc3-4d17-a90e-66b5453935bd",
        "label",
      ),
    ).resolves.toEqual(undefined);

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    await expect(
      inputsValidationStore.validateEntityInput("invalid", "label"),
    ).rejects.toThrowErrorMatchingSnapshot();

    await expect(
      inputsValidationStore.validateEntityInput(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        // @ts-expect-error Intentional wrong data type
        "invalid",
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("can validate a all inputs of a single entity", async () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return z.string().parse(value);
              },
            }),
            createInput({
              name: "maxLength",
              validate(value) {
                return z.number().parse(value);
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {
              // @ts-expect-error Intentional wrong data type
              label: 1,
              // @ts-expect-error Intentional wrong data type
              maxLength: "1",
            },
          },
        },
        root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    await inputsValidationStore.validateEntityInputs(
      "6e0035c3-0d4c-445f-a42b-2d971225447c",
    );

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    await expect(
      inputsValidationStore.validateEntityInputs("invalid"),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("can validate a all inputs of all entities", async () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return z.string().parse(value);
              },
            }),
            createInput({
              name: "maxLength",
              validate(value) {
                return z.number().parse(value);
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {
              // @ts-expect-error Intentional wrong data type
              label: 1,
              maxLength: 1,
            },
          },
          "51324b32-adc3-4d17-a90e-66b5453935bd": {
            type: "test",
            inputs: {
              label: "test",
              // @ts-expect-error Intentional wrong data type
              maxLength: "1",
            },
          },
        },
        root: [
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
          "51324b32-adc3-4d17-a90e-66b5453935bd",
        ],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    await inputsValidationStore.validateEntitiesInputs(),
      expect(inputsValidationStore.getData()).toMatchSnapshot();
  });

  it("can set a single entity input error", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
            createInput({
              name: "title",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {},
          },
        },
        root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    expect(
      inputsValidationStore.setEntityInputError(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "label",
        "Some error",
      ),
    ).toEqual(undefined);

    expect(
      inputsValidationStore.setEntityInputError(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "title",
        "Title error",
      ),
    ).toEqual(undefined);

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(() =>
      inputsValidationStore.setEntityInputError("invalid", "title", "error"),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      inputsValidationStore.setEntityInputError(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        // @ts-expect-error Intentional wrong data type
        "invalid",
        "error",
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it("can set multiple input errors for a single entity", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
            createInput({
              name: "title",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {},
          },
        },
        root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    expect(
      inputsValidationStore.setEntityInputsErrors(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        {
          label: "some error",
          title: "another error",
        },
      ),
    ).toEqual(undefined);

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(() =>
      inputsValidationStore.setEntityInputError("invalid", "title", "error"),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      inputsValidationStore.setEntityInputsErrors(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        {
          // @ts-expect-error Intentional wrong data type
          invalid: "some error",
        },
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it("can set inputs errors for a all entities", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
            createInput({
              name: "title",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {},
          },
          "51324b32-adc3-4d17-a90e-66b5453935bd": {
            type: "test",
            inputs: {},
          },
        },
        root: [
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
          "51324b32-adc3-4d17-a90e-66b5453935bd",
        ],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    expect(
      inputsValidationStore.setEntitiesInputsErrors({
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          label: "some error",
          title: "another error",
        },
        "51324b32-adc3-4d17-a90e-66b5453935bd": {
          label: "some error",
          title: "another error",
        },
      }),
    ).toEqual(undefined);

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(() =>
      inputsValidationStore.setEntitiesInputsErrors({
        invalid: {},
      }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      inputsValidationStore.setEntitiesInputsErrors({
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          // @ts-expect-error Intentional wrong data type
          invalid: "some error",
        },
      }),
    ).toThrowErrorMatchingSnapshot();

    inputsValidationStore.setEntitiesInputsErrors({});

    expect(inputsValidationStore.getData()).toMatchSnapshot();
  });

  it("can set a reset a single entity input error", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "select",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
            createInput({
              name: "title",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "maxLength",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "select",
            inputs: {},
          },
        },
        root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    inputsValidationStore.setEntityInputsErrors(
      "6e0035c3-0d4c-445f-a42b-2d971225447c",
      {
        label: "label error",
        title: "title error",
      },
    );

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(
      inputsValidationStore.resetEntityInputError(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "label",
      ),
    ).toEqual(undefined);

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(() =>
      inputsValidationStore.resetEntityInputError("invalid", "title"),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      inputsValidationStore.resetEntityInputError(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        // @ts-expect-error Intentional wrong data type
        "invalid",
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it("can set a reset all inputs errors for a single entity", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
            createInput({
              name: "title",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {},
          },
        },
        root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    inputsValidationStore.setEntityInputsErrors(
      "6e0035c3-0d4c-445f-a42b-2d971225447c",
      {
        label: "label error",
        title: "title error",
      },
    );

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(
      inputsValidationStore.resetEntityInputsErrors(
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
      ),
    ).toEqual(undefined);

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(() =>
      inputsValidationStore.resetEntityInputsErrors("invalid"),
    ).toThrowErrorMatchingSnapshot();
  });

  it("can set a reset all inputs errors for a all entities", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
            createInput({
              name: "title",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const schemaStore = createSchemaStore({
      builder,
      schema: {
        entities: {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            type: "test",
            inputs: {},
          },
          "51324b32-adc3-4d17-a90e-66b5453935bd": {
            type: "test",
            inputs: {},
          },
        },
        root: [
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
          "51324b32-adc3-4d17-a90e-66b5453935bd",
        ],
      },
    });

    const inputsValidationStore = createInputsValidationStore({
      schemaStore,
    });

    inputsValidationStore.setEntitiesInputsErrors({
      "6e0035c3-0d4c-445f-a42b-2d971225447c": {
        label: "label error",
      },
      "51324b32-adc3-4d17-a90e-66b5453935bd": {
        title: "title error",
      },
    });

    expect(inputsValidationStore.getData()).toMatchSnapshot();

    expect(inputsValidationStore.resetEntitiesInputsErrors()).toEqual(
      undefined,
    );

    expect(inputsValidationStore.getData()).toMatchSnapshot();
  });
});
