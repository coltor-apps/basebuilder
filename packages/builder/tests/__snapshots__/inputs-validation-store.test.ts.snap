// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inputs validation store > can be created 1`] = `
{
  "getData": [Function],
  "getSerializedData": [Function],
  "resetEntitiesInputsErrors": [Function],
  "resetEntityInputError": [Function],
  "resetEntityInputsErrors": [Function],
  "setData": [Function],
  "setEntityInputError": [Function],
  "setEntityInputsErrors": [Function],
  "setSerializedData": [Function],
  "subscribe": [Function],
  "validateEntitiesInputs": [Function],
  "validateEntityInput": [Function],
  "validateEntityInputs": [Function],
}
`;

exports[`inputs validation store > can be created with initial errors 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": [Error],
    },
  },
}
`;

exports[`inputs validation store > can be created with initial errors 2`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can be created with initial errors 3`] = `"Unkown entity input \\"invalid\\"."`;

exports[`inputs validation store > can reset a single entity input error 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": "label error",
      "title": "title error",
    },
  },
}
`;

exports[`inputs validation store > can reset a single entity input error 2`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "title": "title error",
    },
  },
}
`;

exports[`inputs validation store > can reset a single entity input error 3`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can reset a single entity input error 4`] = `"Unkown entity input \\"invalid\\"."`;

exports[`inputs validation store > can reset a single entity input error 5`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "label error",
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "title error",
            "inputName": "title",
          },
        },
      ],
    ],
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": undefined,
            "inputName": "label",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can reset all inputs errors for a single entity 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": "label error",
      "title": "title error",
    },
  },
}
`;

exports[`inputs validation store > can reset all inputs errors for a single entity 2`] = `
{
  "entitiesInputsErrors": Map {},
}
`;

exports[`inputs validation store > can reset all inputs errors for a single entity 3`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can reset all inputs errors for a single entity 4`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "label error",
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "title error",
            "inputName": "title",
          },
        },
      ],
    ],
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": undefined,
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": undefined,
            "inputName": "title",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can reset all inputs errors for all entities 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": "label error",
    },
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "title": "title error",
    },
  },
}
`;

exports[`inputs validation store > can reset all inputs errors for all entities 2`] = `
{
  "entitiesInputsErrors": Map {},
}
`;

exports[`inputs validation store > can reset all inputs errors for all entities 3`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": undefined,
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "51324b32-adc3-4d17-a90e-66b5453935bd",
            "error": undefined,
            "inputName": "title",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can set a single entity input error 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": "Some error",
      "title": "Title error",
    },
  },
}
`;

exports[`inputs validation store > can set a single entity input error 2`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can set a single entity input error 3`] = `"Unkown entity input \\"invalid\\"."`;

exports[`inputs validation store > can set a single entity input error 4`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "Some error",
            "inputName": "label",
          },
        },
      ],
    ],
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "Title error",
            "inputName": "title",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can set multiple input errors for a single entity 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": "some error",
      "title": "another error",
    },
  },
}
`;

exports[`inputs validation store > can set multiple input errors for a single entity 2`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can set multiple input errors for a single entity 3`] = `"Unkown entity input \\"invalid\\"."`;

exports[`inputs validation store > can set multiple input errors for a single entity 4`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "some error",
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": "another error",
            "inputName": "title",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can set raw data 1`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can set raw data 2`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": "some error",
    },
  },
}
`;

exports[`inputs validation store > can set raw data 3`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "DataSet",
          "payload": {
            "data": {
              "entitiesInputsErrors": Map {
                "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
                  "label": "some error",
                },
              },
            },
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can set the data 1`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can set the data 2`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": "some error",
    },
  },
}
`;

exports[`inputs validation store > can set the data 3`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "DataSet",
          "payload": {
            "data": {
              "entitiesInputsErrors": Map {
                "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
                  "label": "some error",
                },
              },
            },
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can validate a all inputs of a single entity 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
      "maxLength": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [],
    "message": "Expected number, received string"
  }
]],
    },
  },
}
`;

exports[`inputs validation store > can validate a all inputs of a single entity 2`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can validate a all inputs of a single entity 3`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [],
    "message": "Expected number, received string"
  }
]],
            "inputName": "maxLength",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can validate a all inputs of all entities 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
    },
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "maxLength": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [],
    "message": "Expected number, received string"
  }
]],
    },
  },
}
`;

exports[`inputs validation store > can validate a all inputs of all entities 2`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": undefined,
            "inputName": "maxLength",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "51324b32-adc3-4d17-a90e-66b5453935bd",
            "error": undefined,
            "inputName": "label",
          },
        },
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "51324b32-adc3-4d17-a90e-66b5453935bd",
            "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [],
    "message": "Expected number, received string"
  }
]],
            "inputName": "maxLength",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`inputs validation store > can validate a single entity input 1`] = `
{
  "entitiesInputsErrors": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
    },
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [],
    "message": "Required"
  }
]],
    },
  },
}
`;

exports[`inputs validation store > can validate a single entity input 2`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`inputs validation store > can validate a single entity input 3`] = `"Unkown entity input \\"invalid\\"."`;

exports[`inputs validation store > can validate a single entity input 4`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
            "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
            "inputName": "label",
          },
        },
      ],
    ],
    [
      [
        {
          "name": "EntityInputErrorUpdated",
          "payload": {
            "entityId": "51324b32-adc3-4d17-a90e-66b5453935bd",
            "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [],
    "message": "Required"
  }
]],
            "inputName": "label",
          },
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
